In [1]:
pip install imbalanced-learn

     |████████████████████████████████| 167 kB 534 kB/s eta 0:00:01
     |████████████████████████████████| 7.2 MB 226 kB/s eta 0:00:01     |██████████▊                     | 2.4 MB 323 kB/s eta 0:00:15
  Attempting uninstall: scikit-learn
    Found existing installation: scikit-learn 0.22.1
    Uninstalling scikit-learn-0.22.1:
      Successfully uninstalled scikit-learn-0.22.1
Note: you may need to restart the kernel to use updated packages.


In [2]:
import pandas as pd
import numpy as np

In [3]:
df=pd.read_csv('creditcard.csv')
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [4]:
df.shape

(284807, 31)

In [5]:
df.columns

Index(['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount',
       'Class'],
      dtype='object')

In [6]:
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [7]:
df['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

In [8]:
X=df.drop('Class', axis=1)
y=df['Class']

In [9]:
from sklearn.linear_model import LogisticRegression

In [11]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

In [12]:
from sklearn.model_selection import KFold

In [13]:
from sklearn.model_selection import GridSearchCV

In [21]:
log_class=LogisticRegression()
grid={'C':10.0**np.arange(-2,3), 'penalty':['l1','l2']}
cv=KFold(n_splits=5, shuffle=False, random_state=None)

In [18]:
from sklearn.model_selection import train_test_split

In [19]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.7)

In [22]:
clf=GridSearchCV(log_class, grid, cv=cv, n_jobs=-1, scoring='f1_macro')
clf.fit(X_train, y_train)

/Users/asap/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


GridSearchCV(cv=KFold(n_splits=5, random_state=None, shuffle=False),
             estimator=LogisticRegression(), n_jobs=-1,
             param_grid={'C': array([1.e-02, 1.e-01, 1.e+00, 1.e+01, 1.e+02]),
                         'penalty': ['l1', 'l2']},
             scoring='f1_macro')

In [23]:
y_pred=clf.predict(X_test)
print (confusion_matrix(y_pred, y_test))
print (classification_report(y_pred, y_test))
print (accuracy_score (y_pred, y_test))

[[198945     94]
 [    90    236]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00    199039
           1       0.72      0.72      0.72       326

    accuracy                           1.00    199365
   macro avg       0.86      0.86      0.86    199365
weighted avg       1.00      1.00      1.00    199365

0.9990770696962857


In [24]:
from sklearn.ensemble import RandomForestClassifier

In [27]:
#it helps to increase the importance of one of the dependent outcomes (fraud 0 and 1 (one is more important than 0))
class_weight=dict({0:1, 1:100})

In [28]:
classifier=RandomForestClassifier(class_weight=class_weight)
classifier.fit(X_train,y_train)

RandomForestClassifier(class_weight={0: 1, 1: 100})

In [29]:
y_pred=classifier.predict(X_test)
print (confusion_matrix(y_pred, y_test))
print (classification_report(y_pred, y_test))
print (accuracy_score (y_pred, y_test))

[[199012     83]
 [    23    247]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00    199095
           1       0.75      0.91      0.82       270

    accuracy                           1.00    199365
   macro avg       0.87      0.96      0.91    199365
weighted avg       1.00      1.00      1.00    199365

0.9994683118902515


#### Under sampling

reduces the points of the maximum labels

it is not really practical to use it, only in cases when we use very small datasets

In [31]:
from imblearn.under_sampling import NearMiss 

In [33]:
y_train.value_counts()

0    85280
1      162
Name: Class, dtype: int64

In [34]:
from collections import Counter
ns=NearMiss(0.8)
X_train_ns, y_train_ns=ns.fit_sample(X_train, y_train)
print("The number of classes before the fit: {}".format(Counter(y_train)))
print("The number of classes after the fit: {}".format(Counter(y_train_ns)))

/Users/asap/anaconda3/lib/python3.7/site-packages/imblearn/utils/_validation.py:638: FutureWarning: Pass sampling_strategy=0.8 as keyword args. From version 0.9 passing these as positional arguments will result in an error
  FutureWarning)


The number of classes before the fit: Counter({0: 85280, 1: 162})
The number of classes after the fit: Counter({0: 202, 1: 162})


It reduced the output to 162=202*0.8

but this dataset is very large and undersampling will have negative effect on accuracy

In [37]:
from sklearn.ensemble import RandomForestClassifier
classifier=RandomForestClassifier()
classifier.fit(X_train_ns,y_train_ns)

RandomForestClassifier()

In [38]:
y_pred=classifier.predict(X_test)
print (confusion_matrix(y_pred, y_test))
print (classification_report(y_pred, y_test))
print (accuracy_score (y_pred, y_test))

[[188482     40]
 [ 10553    290]]
              precision    recall  f1-score   support

           0       0.95      1.00      0.97    188522
           1       0.88      0.03      0.05     10843

    accuracy                           0.95    199365
   macro avg       0.91      0.51      0.51    199365
weighted avg       0.94      0.95      0.92    199365

0.9468663005041005


#### Over sampling

increases the points of the minimum labels

In [40]:
from imblearn.over_sampling import RandomOverSampler

In [42]:
os=RandomOverSampler(0.5)
X_train_os, y_train_os=os.fit_sample(X_train, y_train)
print("The number of classes before the fit: {}".format(Counter(y_train)))
print("The number of classes after the fit: {}".format(Counter(y_train_os)))

The number of classes before the fit: Counter({0: 85280, 1: 162})
The number of classes after the fit: Counter({0: 85280, 1: 42640})


42640=85280*0.5

In [43]:
from sklearn.ensemble import RandomForestClassifier
classifier=RandomForestClassifier()
classifier.fit(X_train_os,y_train_os)

RandomForestClassifier()

In [45]:
y_pred=classifier.predict(X_test)
print (confusion_matrix(y_pred, y_test))
print (classification_report(y_pred, y_test))
print (accuracy_score (y_pred, y_test))

[[199007     78]
 [    28    252]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00    199085
           1       0.76      0.90      0.83       280

    accuracy                           1.00    199365
   macro avg       0.88      0.95      0.91    199365
weighted avg       1.00      1.00      1.00    199365

0.9994683118902515


#### SMOTETomek

it uses both under sampling and over sampling

In [46]:
from imblearn.combine import SMOTETomek

In [48]:
Smote=SMOTETomek(0.75)
X_train_smote, y_train_smote=Smote.fit_sample(X_train, y_train)
print("The number of classes before the fit: {}".format(Counter(y_train)))
print("The number of classes after the fit: {}".format(Counter(y_train_smote)))

/Users/asap/anaconda3/lib/python3.7/site-packages/imblearn/utils/_validation.py:638: FutureWarning: Pass sampling_strategy=0.75 as keyword args. From version 0.9 passing these as positional arguments will result in an error
  FutureWarning)


The number of classes before the fit: Counter({0: 85280, 1: 162})
The number of classes after the fit: Counter({0: 84612, 1: 63292})


In [49]:
from sklearn.ensemble import RandomForestClassifier
classifier=RandomForestClassifier()
classifier.fit(X_train_smote,y_train_smote)

RandomForestClassifier()

In [50]:
y_pred=classifier.predict(X_test)
print (confusion_matrix(y_pred, y_test))
print (classification_report(y_pred, y_test))
print (accuracy_score (y_pred, y_test))

[[198988     69]
 [    47    261]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00    199057
           1       0.79      0.85      0.82       308

    accuracy                           1.00    199365
   macro avg       0.90      0.92      0.91    199365
weighted avg       1.00      1.00      1.00    199365

0.9994181526346149


#### Ensemble Technique

In [51]:
from imblearn.ensemble import EasyEnsembleClassifier

In [53]:
easy=EasyEnsembleClassifier()
easy.fit(X_train, y_train)

EasyEnsembleClassifier()

In [54]:
y_pred=classifier.predict(X_test)
print (confusion_matrix(y_pred, y_test))
print (classification_report(y_pred, y_test))
print (accuracy_score (y_pred, y_test))

[[198988     69]
 [    47    261]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00    199057
           1       0.79      0.85      0.82       308

    accuracy                           1.00    199365
   macro avg       0.90      0.92      0.91    199365
weighted avg       1.00      1.00      1.00    199365

0.9994181526346149
